File: dsa/helltime1.0/binary/1482.cpp
class Solution
{
public:
    int minDays(vector<int> &bloomDay, int m, int k)
    {
        if ((long long)m * k > bloomDay.size())
        {
            return -1;
        }
        int lower = *min_element(bloomDay.begin(), bloomDay.end());
        int upper = *max_element(bloomDay.begin(), bloomDay.end());
        function<bool(int)> can_make = [&](int capacity) -> bool
        {
            int flowers = 0;
            int bouquets = 0;
            for (const int &bloom : bloomDay)
            {
                flowers = (bloom <= capacity) ? (flowers + 1) : 0;
                if (flowers == k)
                {
                    flowers = 0;
                    bouquets++;
                }
            }
            return bouquets >= m;
        };
        while (lower < upper)
        {
            int mid = (lower + (upper - lower) / 2);
            if (can_make(mid))
            {
                upper = mid;
            }
            else
            {
                lower = mid + 1;
            }
        }
        return lower;
    }
};

File: dsa/helltime1.0/extras2.0/1482.py
class Solution:
    def minDays(self, bloomDay: List[int], m: int, k: int) -> int:
        
        def process(mid):
            bouquests = 0
            cnt = 0
            for day in bloomDay:
                if day <= mid:
                    cnt += 1
                else:
                    cnt = 0

                if cnt == k:
                    bouquests += 1
                    cnt = 0
            
            return bouquests

        N = len(bloomDay)
        if m * k > N:
            return -1

        start = 0
        end = max(bloomDay)
        minDays = -1

        while start <= end:
            mid = (start + end) // 2
            if  process(mid) >= m:
                minDays = mid
                end = mid - 1
            else:
                start = mid + 1
        
        return minDays
